samples/bluetooth/bap_broadcast_source/src/main.c

```diff
@@ -303,24 +315,59 @@ static void init_lc3_thread(void *arg1, void *arg2, void *arg3)
 K_THREAD_DEFINE(encoder, LC3_ENCODER_STACK_SIZE, init_lc3_thread, NULL, NULL, NULL,
 		LC3_ENCODER_PRIORITY, 0, -1);
 
-#if defined(CONFIG_USB_DEVICE_AUDIO)
-static void data_received(const struct device *dev, struct net_buf *buffer, size_t size)
+#if defined(CONFIG_USE_USB_AUDIO_INPUT)
+/* Allocate 3: 1 for USB to receive data to and 2 additional buffers to prevent out of memory
+ * errors when USB host decides to perform rapid terminal enable/disable cycles.
+ */
+K_MEM_SLAB_DEFINE_STATIC(usb_in_buf_pool, USB_MAX_STEREO_FRAME_SIZE, 3, 4);
+static bool terminal_enabled;
+
+static void terminal_update_cb(const struct device *dev, uint8_t terminal, bool enabled,
+			       bool microframes, void *user_data)
 {
-	static int count;
-	int16_t *pcm;
-	int nsamples, ratio;
-	int16_t usb_pcm_data[USB_CHANNELS][USB_NUM_SAMPLES];
+	terminal_enabled = enabled;
+}
 
-	if (!buffer) {
-		return;
+static void uac2_sof_cb(const struct device *dev, void *user_data)
+{
+	/* no-op, but has to be registered */
+}
+
+static void *get_recv_buf_cb(const struct device *dev, uint8_t terminal, uint16_t size,
+			     void *user_data)
+{
+	void *buf = NULL;
+	int ret;
+
+	if (!terminal_enabled) {
+		return NULL;
+	}
+
+	__ASSERT(size <= USB_MAX_STEREO_FRAME_SIZE, "%u was not <= %d", size,
+		 USB_MAX_STEREO_FRAME_SIZE);
+
+	ret = k_mem_slab_alloc(&usb_in_buf_pool, &buf, K_NO_WAIT);
+	if (ret != 0) {
+		printk("Failed to allocate buffer: %d\n", ret);
 	}
 
-	if (!size) {
-		net_buf_unref(buffer);
+	return buf;
+}
+
+static void data_recv_cb(const struct device *dev, uint8_t terminal, void *buf, uint16_t size,
+			 void *user_data)
+{
+	int16_t usb_pcm_data[USB_CHANNELS][USB_DOWNSSAMPLE_CNT];
+	int nsamples, ratio;
+	static int count;
+	int16_t *pcm;
+
+	if (!terminal_enabled || buf == NULL || size == 0U) {
+		k_mem_slab_free(&usb_in_buf_pool, buf);
 		return;
 	}
 
-	pcm = (int16_t *)net_buf_pull_mem(buffer, size);
+	pcm = (int16_t *)buf;
 
 	/* 'size' is in bytes, containing 1ms, 48kHz, stereo, 2 bytes per sample.
 	 * Take left channel and do a simple downsample to 16kHz/24Khz
```

Comment:
As it's supporting stereo, maybe you could fix a typo here (made by me originally ... should not be 'left' :))